Add more descriptions & examples to the readme (#316)
canonical/microcluster: doc/clustering.md

# Initialize a Microcluster

> ## Bootstrap the First Member

Reviewer: ```suggestion
## Bootstrap the first member
```

> The first member can be bootstrapped with a request on the local unix socket using

Reviewer: ```suggestion
The first member can be bootstrapped with a request on the local unix socket using:
```

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}

// Cluster member name, usually hostname.
memberName := "member1"

// Cluster member address and port.
memberAddress := "10.0.0.101:8000"

// Config to pass into PreInit and PostBootstrap lifecycle hooks.
cfg := map[string]string{"key": "val"}

// Bootstrap the cluster member.
err = m.NewCluster(ctx, memberName, memberAddress, cfg)
if err != nil {
    // ...
}
```

> ## Issue a Join Token

Reviewer: ```suggestion
## Issue a join token
```

> A join token can be issued to allow a new cluster member with a given name to join the cluster. A join token is a `base64` encoded string that must be supplied with a join request.

Reviewer: ```suggestion
A join token allows a new cluster member with a given name to join the cluster. A join token is a `base64` encoded string that must be supplied with a join request.
```

> Join tokens can only be issued over the local unix socket of an initialized cluster member. To issue a token, a cluster member name and token expiry must be provided. The given name must be present in the `server.crt` Subject Alternative Names of the joining cluster member.

Reviewer: ```suggestion
Join tokens can only be issued over the local unix socket of an initialized cluster member. To issue a token, a cluster member name and token expiry must be provided. The given name must be present in the `server.crt` Subject Alternative Names (SAN) of the joining cluster member.
```

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}

// Cluster member name of the joiner. It will be verified against the server.crt SAN of the joiner when the token is used.
joinerName := "member2"

// How long the join token will be valid before its record is deleted.
expireAfter := 3 * time.Hour

// Issue the join token.
token, err := m.NewJoinToken(ctx, joinerName, expireAfter)
if err != nil {
    // ...
}
```

> ## Join an Existing Cluster with a Join Token

Reviewer: ```suggestion
## Join an existing cluster with a join token
```

Use an existing join token to add a new member to the cluster. This request can only be initiated over the local unix socket of the joiner.

Before triggering the [PostJoin](https://github.com/canonical/microcluster/blob/4d80df396e335bf26f9895956e846e082bb8f624/internal/state/hooks.go#L23) lifecycle hook, all previously existing cluster members will concurrently run their [OnNewMember](https://github.com/canonical/microcluster/blob/4d80df396e335bf26f9895956e846e082bb8f624/internal/state/hooks.go#L39) lifecycle hooks.

```go
m, err := microcluster.App(microcluster.Args{StateDir: "/path/to/state"})
if err != nil {
    // ...
}

// Cluster member name, usually hostname. This must match the name provided when the join token was issued.
memberName := "member2"

// Cluster member address and port.
memberAddress := "10.0.0.102:8000"

// Config to pass into PreInit and PostJoin lifecycle hooks.
cfg := map[string]string{"key": "val"}

// Request to join a cluster with the provided join token.
err = m.JoinCluster(ctx, memberName, memberAddress, token, cfg)
if err != nil {
    // ...
}
```

> ## Remove a Cluster Member

Reviewer: ```suggestion
## Remove a cluster member
```
